pkg/sql/descmetadata/metadata_updater.go

```diff
@@ -184,3 +188,39 @@ func (mu metadataUpdater) DeleteSchedule(ctx context.Context, scheduleID int64)
 	)
 	return err
 }
+
+// SetZoneConfig implements scexec.DescriptorMetadataUpdater.
+func (mu metadataUpdater) SetZoneConfig(
+	ctx context.Context, id descpb.ID, zone *zonepb.ZoneConfig,
+) error {
+	if zone == nil {
+		_, err := mu.ie.Exec(ctx, "delete-zone", mu.txn,
+			"DELETE FROM system.zones WHERE id = $1", id)
+		return err
+	}
+
+	return mu.collectionFactory.Txn(ctx, mu.ie, mu.txn.DB(),
+		func(ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {
+
+			// Recompute any spans for the subzones.
+			if len(zone.Subzones) != 0 {
+				tbl, err := descriptors.GetImmutableTableByID(ctx, txn, id, tree.ObjectLookupFlagsWithRequiredTableKind(tree.ResolveRequireTableDesc))
+				if err != nil {
+					return err
+				}
+				zone.SubzoneSpans, err = regionutils.GenerateSubzoneSpans(mu.codec, tbl, zone.Subzones)
+				if err != nil {
+					return err
+				}
+			}
+
+			bytes, err := protoutil.Marshal(zone)
+			if err != nil {
+				return err
+			}
+			_, err = mu.ie.Exec(ctx, "update-zone", mu.txn,
+				"UPSERT INTO system.zones (id, config) VALUES ($1, $2)", id, bytes)
+			return err
+		})
```

Comment:
This mixing of txns is suspect. I see it being done above, but it all makes me uneasy. A simple-enough refactoring here would be to make the `metadataUpdater` not free-standing, but instead have the MetadataUpdaterFactory take a descs.Collection and kv.Txn when being constructed. It already takes a txn. 

Honestly, the `scexec.DescriptorMetadataUpdaterFactory` ought not be defined in `scexec` but rather in `scdeps`. Then it can reference the `*descs.Collection` which is already in scope. WDYT?

I think this refactoring can be pulled out to a separate PR.

Reply:
Yes, I'll split this off and get a PR out